### PR TITLE
feat: support tags/hashtags on examples to improve discovery #422

### LINKED ISSUE #422: Support tags/hashtags on examples to improve discovery
The number of component-specific examples is going to increase over the next few releases.

- Support hashtags in the example description headers.
- Create an index page that lists all examples grouped by tag.

Needs changes to https://github.com/h2oai/wave/blob/master/py/sync_examples.py

### PR DESCRIPTION
**Pull request by srini-x**
Closes #422 

1. Added a new page `Tags` between `Gallery` and `Hello World!`
2. The `tags` are subheadings on this page, sorted alphabetically. Under each `tag`, links to the examples with those tags are available.
3. Added a `Tags` field under each example with tags. These tags are links that link back to the Tags page mentioned above.
4. Added `keywords` metadata in the header of the each generated `.md` file. Docusaurus [markdown headers](https://v2.docusaurus.io/docs/markdown-features#markdown-headers) provide only `keywords` under Docs and `tags` under Blog.
5. In the python example files, tags can be provided as the last line of description.
6. Tags line can not have `.` in it. Tags must be `,` separated. A tag can not contain spaces.

ToDo:
1. Comment the code
2. Add tags to all remaining examples.

@lo5, If you like this implementation, I will go ahead and complete the ToDos listed above.

example:

`buttons.py`

```py
# Form / Buttons
# Use the `ui.buttons()` function to group related buttons.
# form, buttons
# ---
from h2o_wave import main, app, Q, ui
```

`buttons.md`

```md
---
title: Form / Buttons
keywords:
  - form
  - buttons
---
```

wave_docs_examples_tags

wave_docs_examples_buttons_tags

### CODE REVIEW

**Review by lo5 — CHANGES_REQUESTED** [review]
@srini-x This is a great idea - thanks for the PR.

It would be good for `get_keywords_if_available()` to be a `parse_tags(description: str)` that accepts a description containing tags and returns a `(new_description, tags)` tuple.

The convention for tags can be `#tag`. For example, accept a description 
```
Use a checklist to group a set of related checkboxes. #form #checkbox #checklist
```
or
```
Use a checklist to group a set of related checkboxes. 
#form #checkbox #checklist
```

or

```
Use a #checklist to group a set of related checkboxes. 
#form #checkbox
```

... and return `('Use a checklist to group a set of related checkboxes.', ['form', 'checkbox', 'checklist'])`

This way, tags can be optional, free-form, and we don't have to stick to a convention that the last line is reserved for tags, which might seem odd when one is staring at the example's header in isolation in an editor.

**Comment by lo5** [comment]
@srini-x - Also, this has minor conflicts with @dulajra 's change. Let's name the new page "Examples by Tag" and add it after the "All Examples" page.

**Comment by srini-x** [comment]
@lo5

- implemented `parse_tags` function
- added tags for all examples
- double checked for typos and broken tags

wave_examples_by_tag

File: py/sync_examples.py [hunk 1]
```diff
@@ -62,15 +91,82 @@ def write_file(p: str, txt: str) -> str:
...
+    pattern = re.compile(hashtag_regex_pattern)
+    matches = pattern.findall(' ' + description)
+
+    # Retrieve tags from the matches
+    tags = sorted(list(set([x[-1].lower() for x in matches])))
+
+    # Remove the '#' before the tags in description
+    new_d = pattern.sub(r'\1\2', ' ' + description)
+
+    # Remove the last line in description if it has only tags
```
Comment (lo5) [inline py/sync_examples.py:142]: Not sure why this part is required. We could simply trim the description.

File: py/sync_examples.py [hunk 2]
```diff
@@ -98,6 +194,30 @@ def make_gallery(examples: List[Example]):
...
+    example_links = [f"- <a href={{useBaseUrl('docs/examples/{e.slug}')}}>{e.title}</a>" for e in examples]
+    return f'### {tag}\n' + '\n' + '\n'.join(example_links) + '\n'
+
+
+def index_examples(examples: List[Example]):
+    tags_dict = defaultdict(list)
+    for e in examples:
+        for t in e.tags:
+            tags_dict[t].append(e)
+    return sorted(tags_dict.items())
```
Comment (lo5) [inline py/sync_examples.py:207]: Need to sort by the example's title.

**Review by lo5 — CHANGES_REQUESTED** [review]
Thanks for the PR @srini-x.

The following tags look overly granular to me, and can be removed - most of these have one example under them, which makes the tags page less useful:
- application
- area
- async
- axis_title
- backend
- bar
- bubble_plot
- clock
- color
- column
- combined
- curve
- custom
- d3
- data
- dynamic
- fill
- filter
- font_size
- game_of_life
- gantt
- gauge
- glider_gun
- global_variables
- group_by
- grouped
- heatmap
- helix
- hilbert
- issue_tracker
- large
- layer
- line
- live
- multi_user
- multiple_categories
- occlusion
- opacity
- path
- point
- polar
- polygon
- position
- primitives
- racetrack
- realtime
- region
- rose
- scatterplot
- shape
- simple
- size
- small
- smooth
- sort
- spline
- state
- status
- step
- step_after
- step_before
- stroke
- stroked
- tall
- theta
- title
- transpose
- turtle
- update
- wide

Additionally:
- labels -> merge plural with singular
- preselection -> merge selection
- vega_lite -> merge with vega

File: py/sync_examples.py (hunk 2 above)
Comment (srini-x) [reply py/sync_examples.py:207]: @lo5 `label` is for `ui.label` and `labels` is for data labels in plots.

File: py/sync_examples.py (hunk 1 above)
Comment (srini-x) [reply py/sync_examples.py:142]: @lo5 This is to make this example work.

```
Use a #checklist to group a set of related checkboxes. 
#form #checkbox
```

If the tags are part of the description, we don't want to remove the entire word, but only the `#` in front of it.

I trim the description, if the tags are at the end of it.

**Comment by srini-x** [comment]
@lo5 

fixed the following

- [x] Removed the overly granular tags
- [x] Merged `labels` -> `label`
- [x] Merged `preselection` -> `selection`
- [x] Merged `vega_lite` -> `vega`
- [x] Merged `nav` -> `navigation`
- [x] Examples within each tag are sorted

File: py/sync_examples.py (hunk 2 above)
Comment (lo5) [reply py/sync_examples.py]: Yes, in that case let's drop `labels`. Thanks @srini-x.

**Comment by srini-x** [comment]
@lo5 merged commits from master

File: py/examples/table.py [hunk 3]
```diff
@@ -1,9 +1,11 @@
 # Table
-# Use a table to display tabular data.
+# Use a #table to display tabular data.
 # ---
 import random
+
 from faker import Faker
-from h2o_wave import main, app, Q, ui
+
+from h2o_wave import app, Q, ui
```
Comment (lo5) [inline py/examples/table.py:8]: The `main` import is required here.

**Comment by srini-x** [comment]
@lo5 removed `labels` tag.

**Review by lo5 — CHANGES_REQUESTED** [review]
Thanks @srini-x - looks good. We'll merge after the correction to `py/examples/table.py`.

File: py/examples/table.py (hunk 3 above)
Comment (srini-x) [reply py/examples/table.py]: fixed.

**Review by lo5 — APPROVED** [review]
Thanks @srini-x!

**Pull request merged**